docs/adrs/0012-exposure-events-and-coenrolling-features.md

```diff
@@ -0,0 +1,142 @@
+# Exposure events and coenrolling features
+
+* Status: proposed
+* Deciders: dberry@mozilla.com, jhugman@mozilla.com, jlockhart@mozilla.com, shell@mozilla.com
+* Date: 2023-07-04, making the decision on or before 2023-07-11.
+
+Technical Story: [Supporting exposure events for coenrolled features](https://mozilla-hub.atlassian.net/browse/EXP-3602)
+
+## Context and Problem Statement
+
+Application feature code can be configured from and by Nimbus. In the case of `messaging` and `onboarding`, the complete set of messages and all their associated data is implemented by this feature configuration.
+
+The feature configuration is composed by merging together the default configuration, any configuration from a rollout, and any configuration from experiments.
```

Comment:
I thought that:
If the default configuration for a feature and no experiment = default config
if a user is enrolled in an experiment - it uses the experiment values + default values for any not defined.
If a user is enrolled in a rollout - it uses the rollout values + default values for any not defined.
If a user is enrolled in an experiment AND a rollout - it applies all the experiment values + default values for any not applied.

Reply:
Does coenrollment apply to both messaging experiments and rollouts?

Reply:
You're right with the rules you've given. The `merge` operation is defined by (pseudo code):

```
fn merge(override: JSONValue, fallback: JSONValue) -> JSONValue:
	when override is None => return fallback,
	when override is JSONObject, fallback is JSONObject =>
			result = clone(fallback)
			foreach key in override.keys():
				result[key] = merge(override[key], fallback[key])
			return result
	otherwise => return override
```

Multiple merges happens, depending on the feature and the experiments currently in play.

For completeness/documentation:

If the default configuration for a feature is:
```json
{
  "a": "default",
  "b": "default",
  "c": "default",
  "M": {}
}
```

and a experiment varies: `"a"`, and adds an `exp` to the `M` object:

```json
{
  "a": "experiment",
  "M": {
	  "exp": 1
  }
}
```

and a rolliout varies "a" and "b", and adds `ro` to the `M` object:

```json
{
   "a": "rollout",
   "b": "rollout",
   "M": {
	  "ro": 2
   }
}
```

Then a user in the experiment gets: 
- Experiment + Defaults, i.e. `merge(experiment, defaults)`
```json
{
  "a": "experiment",
  "b": "default",
  "c": "default",
  "M": {
    "exp": 1
  }
}
```


For a user in the rollout, but not the experiment:
- Rollout + Defaults, i.e. `merge(rollout, defaults)`
```json
{
  "a": "rollout",
  "b": "rollout",
  "c": "default",
  "M": {
    "ro": 2
  }
}
```

And a user in both rollout and experiment gets:
Experiment + Rollout + Defaults, i.e. `merge(experiment, merge(rollout, defaults))`
```json
{
  "a": "experiment",
  "b": "rollout",
  "c": "default",
  "M": {
    "exp": 1,
    "ro": 2,
  }
}
```

For coenrolling features, a feature can be involved in more than one experiment, so the feature config can be:

Experiment 1 + Experiment2 + Rollout1 + Rollout2 + Defaults i.e. 

So if each of the experiments and rollouts contribute something with a different key in `M` , for example:

experiment1
```json
{ "M": { "exp1": 1 } }
```

experiment2
```json
{ "M": { "exp2": 2 } }
```

rollout1
```json
{ "M": { "ro1": 3 } }
```

rollout2
```json
{ "M": { "ro1": 4 } }
```

`merge(merge(merge(experiment1, experiment2), merge(rollout1, rollout2), defaults)` gives:
```json
{
  "a": "default",
  "b": "default",
  "c": "default",
  "M": {
    "exp1": 1,
    "exp2": 2,
    "ro1": 3,
    "ro2": 4
  }
}
```

The ADR asks how do we unpick where does the `2` come from? For `messaging`, the `M` is `messages`, and the integers which are values are objects.

So for `experiment1` may be:
```json
{
	"messages": {
		"my-message-1": {
			"experiment": "{experiment}"
		}
	}
}
```

and `experiment2` may be:
```json
{
	"messages": {
		"my-message-2": {
			"experiment": "{experiment}"
		}
	}
}
```

and it the `{experiment}` that the user typed gets replaced, and merged into:

```json
{
	"messages": {
		"my-message-1": {
			"experiment": "experiment1",
		},
		"my-message-2": {
			"experiment": "experiment2",
		},
	}
}
```

When `my-message-1` is displayed, then we can get the `experiment` property, and find out that it was the experiment with slug `experiment1`  that contributed this message.

We can go further, and replace `{experiment}` in the message key, so the user types in *both* `experiment1` and `experiment2`:

```json
{
	"messages": {
		"{experiment}-en": {
			"experiment": "{experiment}"
		},
		"{experiment}-es": {
			"experiment": "{experiment}"
		},
		"{experiment}-fr": {
			"experiment": "{experiment}"
		}
	}
}
```
and the replace/merge yields:

```json
{
	"messages": {
		"experiment1-en": {
			"experiment": "experiment1"
		},
		"experiment1-es": {
			"experiment": "experiment1"
		},
		"experiment1-fr": {
			"experiment": "experiment1"
		},
		"experiment2-en": {
			"experiment": "experiment2"
		},
		"experiment2-es": {
			"experiment": "experiment2"
		},
		"experiment2-fr": {
			"experiment": "experiment2"
		}
	}
}
```